src/Audio/Audio.man
.\" t
.\" @(#)Audio.1	02/07/00
.TH Audio 1.8 "Feb 7 2000"
.UC
.SH NAME
\fBAudio\fP \- the AfterStep Sounds module

.SH SYNOPSIS
\fBAudio\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
.sp
Although \fBAudio\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
	Module	"I"	Audio 
.sp
.nf
or from within an afterstep menu file by 'Audio':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Audio"
	Module "Audio" Audio
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBAudio\fP module communicates with the AfterStep window manager to
bind audio sounds to window manager actions. Different audio queues may be
assigned to any window manager action. \fBAudio\fP can be used with
any independent audio player, and therefore requires no special audio
library APIs of its own. Simple in its design, it merely binds the audio
filenames to particular actions and forks off the audio player program
with the appropriate filename when that action occurs. 

If You have Rplay client-server installed and working \fBAudio\fP can use 
libRplay. It will try to autodetect it at the time of running configure
and if it finds it  - "builtin-rplay" will be placed as the default
AudioPlay type.

In addition Audio can copy Audio samples from .au files into /dev/audio
on itself, or it can dump contents of the specified files into STDOUT, 
so you can redirect it anywhere you want.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fBAudio\fP will search for a configuration file
which describes the audio player and sound files associated with actions,
to use.  This file will be the one specified by the \fI\-f\fP option (if 
given).  Else, the ~/GNUstep/Library/AfterStep/audio file will be used 
if it exists.  Else, /usr/local/share/afterstep/audio will be used.

.SH INVOCATION
\fBAudio\fP can be invoked from the \fIautoexec\fP file or from a menu
item as shown in the Synopsis section above.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate \fBAudio\fP.

.SH CONFIGURATION OPTIONS
\fBAudio\fP reads the \fIaudio\fP file when it starts up and looks for
its configuration options.

.IP "*AudioPath \fIpath\fP"
Specifies the \fIpath\fP to look for the audio files.  This option
is ignored when rplay is used.

.IP "*AudioPlayCmd \fIplayer\fP"
This determines parameters for audio player program that will actually
play the sounds, or any of the built-in players.  
The following are built-in options :
.IP ""
.nf
\fIbuiltin-rplay\fP -  builtin rplay support will be used 
(must have rplay installed);
\fIbuiltin-cat\fP - sound data will be sent to /dev/audio 
directly (this can only be used if AfterStep runs on
the same computer as X Server).
\fIbuiltin-stdout\fP - sound data will be dumped into 
STDOUT so you can redirect it anywhere you want.
.fi
.IP ""
Anything other then these built in options will be considered a command 
line parameters for the external player. External player itself has to 
be specified at compile time using --with-soundplayer option for configure.
Default is "play" utility from the SoX package.
If this option is ommited then external player is used with no parameters.
.IP ""
Note: sound file to be played always goes as the last parameter in external 
player's command line.

.IP "*AudioDelay \fIseconds\fP"
Specifies that sound events will only be played if they occur
at least so many \fIseconds\fP after the previous event.  Sound events that
occur during the delay period are ignored.  This option is useful if 
you don't want several sounds playing at the same time.  The default
delay is 0 which disables the audio delay.

.IP "*AudioRplayHost \fIhostname\fP"
Specifies what host the sounds will play on.  The \fIhostname\fP can also
be an environment variable such as $HOSTDISPLAY. This option is only valid
with builtin rplay support. If this is left out then Audio will attempt
to connect to rplay daemon on the same host that is running X first, then
to localhost. 

.IP "*AudioRplayPriority \fIpriority\fP"
Specifies what \fIpriority\fP will be assigned to the sounds when they
are played.  This option is only valid with builtin rplay support.

.IP "*AudioRplayVolume \fIlevel\fP"
Specifies what volume \fIlevel\fP will be assigned to the sounds when they
are played.  This option is only valid with builtin rplay support.

.IP "*Audio \fIaction filename\fP"
Binds the particular window manager \fIaction\fP that the sound
\fIfilename\fP is played for.

.RE
.PP
.RS
The following \fBaction\fPs are available for sounds:
.IP ""
.nf
startup 
shutdown
unknown
add_window
raise_window
lower_window
focus_change
destroy_window
iconify
deiconify
toggle_paging
new_page
new_desk
configure_window
window_name
icon_name
res_class
res_name
end_windowlist
.fi

.SH AUTHORS
Sasha Vasko <sashav at sprintmail.com>
.nf
Frank Fejes <frank@ssax.com>
.nf
Dan Weeks <dan@mango.sfasu.edu>
.nf
Szijarto Szabolcs <saby@sch.bme.hu>
.nf
Mark Boyns <boyns@sdsu.edu>
.nf
Mark Scott <mscott@mcd.mot.com>
.nf
Robert Nation
.nf
David Mihm <dmihm@swbell.net> [man page]

.SH SEE ALSO
.BR afterstep (1),
.BR rplay (1),
.BR showaudio (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

libAfterConf/Audio.c
/*
 * Copyright (c) 2000 Andrew Ferguson <andrew@owsla.cjb.net>
 * Copyright (c) 1998 Sasha Vasko <sasha at aftercode.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */


#define LOCAL_DEBUG

#include "../configure.h"

#include "../libAfterStep/asapp.h"
#include "../libAfterStep/afterstep.h"
#include "../libAfterStep/parser.h"
#include "../libAfterStep/hints.h"
#ifdef HAVE_RPLAY_H
#include <rplay.h>
#endif

#include "afterconf.h"

/*****************************************************************************
 *
 * This routine is responsible for reading and parsing the config file
 *
 ****************************************************************************/

TermDef       EventTerms[] = {
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "add_window", 10,    	TT_FILENAME, EVENT_WindowAdded_ID        , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "window_names", 8,      TT_FILENAME, EVENT_WindowNames_ID        , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "destroy_window", 14,   TT_FILENAME, EVENT_WindowDestroyed_ID    , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "focus_change", 12,	    TT_FILENAME, EVENT_WindowActivated_ID    , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "raise_window", 12,  	TT_FILENAME, EVENT_WindowRaised_ID       , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "iconify", 7,     		TT_FILENAME, EVENT_WindowIconified_ID    , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "deiconify", 9,    		TT_FILENAME, EVENT_WindowDeiconified_ID  , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "shade", 5,     		TT_FILENAME, EVENT_WindowShaded_ID       , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "unshade", 7,      		TT_FILENAME, EVENT_WindowUnshaded_ID     , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "stick", 5, 			TT_FILENAME, EVENT_WindowStuck_ID        , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "unstick", 7,   		TT_FILENAME, EVENT_WindowUnstuck_ID      , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "maximize", 8, 			TT_FILENAME, EVENT_WindowMaximized_ID    , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "restore", 7,   		TT_FILENAME, EVENT_WindowRestored_ID     , NULL},
    
	{TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "new_background", 14,   TT_FILENAME, EVENT_BackgroundChanged_ID  , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "new_viewport", 12,		TT_FILENAME, EVENT_DeskViewportChanged_ID, NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "startup", 7,           TT_FILENAME, EVENT_Startup_ID            , NULL},
    {TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "shutdown", 8,          TT_FILENAME, EVENT_Shutdown_ID           , NULL},
    
	{TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "config", 6,         	TT_FILENAME, EVENT_Config_ID             , NULL},
	{TF_DONT_SPLIT|TF_NO_MYNAME_PREPENDING, "module_config", 13,    TT_FILENAME, EVENT_ModuleConfig_ID       , NULL},
    
	{0, NULL, 0, 0, 0}						   /* end of structure */
};

SyntaxDef     AudioEventsSyntax = {
	',',
    '\n',
    EventTerms,
	0,										   /* use default hash size */
	' ',
	"",
	"\t",
    "Audio event configuration",
	"AudioEvents",
	"names for different windowing events",
	NULL,
	0
};


TermDef       AudioTerms[] = {
    {TF_DONT_SPLIT, "PlayCmd", 7, TT_TEXT, AUDIO_PLAYCMD_ID, NULL},
    {0, "Delay", 5, TT_INTEGER, AUDIO_DELAY_ID, NULL},
    {0, "RplayHost", 9, TT_TEXT, AUDIO_RPLAY_HOST_ID, NULL},
    {0, "RplayPriority", 13, TT_INTEGER, AUDIO_RPLAY_PRI_ID, NULL},
    {0, "RplayVolume", 11, TT_INTEGER, AUDIO_RPLAY_VOL_ID, NULL},
    {0, "", 0, TT_FLAG, AUDIO_AUDIO_ID, &AudioEventsSyntax},
	{0, NULL, 0, 0, 0}						   /* end of structure */

};

SyntaxDef     AudioSyntax = {
	'\n',
	'\0',
	AudioTerms,
	0,										   /* use default hash size */
	' ',
	"",
	"\t",
	"Audio configuration",
	"Audio",
	"AfterStep module for playing sounds when windowing event occurs",
	NULL,
	0
};

/************** Create/Destroy AudioConfig ***************/

AudioConfig  *
CreateAudioConfig ()
{
    AudioConfig  *config = (AudioConfig *) safecalloc (1, sizeof (AudioConfig));

    /* let's initialize Audio's config with some nice values: */
#ifdef HAVE_RPLAY_H
    config->rplay_volume = RPLAY_DEFAULT_VOLUME;
    config->rplay_priority = RPLAY_DEFAULT_PRIORITY;
#endif

	return config;
}

void
DestroyAudioConfig (AudioConfig * config)
{
    register int i ;

   	for( i = EVENT_ID_END-EVENT_ID_START-1 ; i >=0 ; i-- )
       	  if( config->sounds[i] )
              free( config->sounds[i] );

	if (config->playcmd)
		free (config->playcmd);
	if (config->rplay_host)
		free (config->rplay_host);
	DestroyFreeStorage (&(config->more_stuff));
	free (config);
}

/************ Actual Parsing **************/

AudioConfig  *
ParseAudioOptions (const char *filename, char *myname)
{
	ConfigData cd ;
	ConfigDef    *AudioConfigReader;
	AudioConfig  *config = CreateAudioConfig ();

	FreeStorageElem *Storage = NULL, *pCurr;
	ConfigItem    item;

	cd.filename = filename ;
	AudioConfigReader = InitConfigReader (myname, &AudioSyntax, CDT_Filename, cd, NULL);
	if (!AudioConfigReader)
		return config;

	item.memory = NULL;
	PrintConfigReader (AudioConfigReader);
	ParseConfig (AudioConfigReader, &Storage);
	PrintFreeStorage (Storage);

	/* getting rid of all the crap first */
	StorageCleanUp (&Storage, &(config->more_stuff), CF_DISABLED_OPTION);

	for (pCurr = Storage; pCurr; pCurr = pCurr->next)
	{
		if (pCurr->term == NULL)
			continue;

		{
			if (!ReadConfigItem (&item, pCurr))
				continue;

			switch (pCurr->term->id)
			{
			 case AUDIO_AUDIO_ID:
                 if( pCurr->sub )
                 {
                    if( pCurr->sub->term && pCurr->sub->argv )
                        if( pCurr->sub->term->id >= EVENT_ID_START && pCurr->sub->term->id < EVENT_ID_END )
                            set_string_value( &(config->sounds[pCurr->sub->term->id-EVENT_ID_START]), mystrdup( pCurr->sub->argv[0] ), NULL, 0 );
                 }
				 break;
			 case AUDIO_PLAYCMD_ID:
                 set_string_value( &(config->playcmd), item.data.string, NULL, 0 );
				 break;
			 case AUDIO_DELAY_ID:
			 	 set_flags( config->set_flags, AUDIO_SET_DELAY );
				 config->delay = (int)item.data.integer;
				 break;
			 case AUDIO_RPLAY_HOST_ID:
                 set_string_value( &(config->rplay_host), item.data.string, &(config->set_flags), AUDIO_SET_RPLAY_HOST );
				 break;
			 case AUDIO_RPLAY_PRI_ID:
				 set_flags( config->set_flags, AUDIO_SET_RPLAY_PRIORITY );
				 config->rplay_priority = (int)item.data.integer;
				 break;
			 case AUDIO_RPLAY_VOL_ID:
				 set_flags( config->set_flags, AUDIO_SET_RPLAY_VOLUME );
				 config->rplay_volume = (int)item.data.integer;
				 break;
			 default:
				 item.ok_to_free = 1;
			}
		}
	}

	ReadConfigItem (&item, NULL);

	DestroyConfig (AudioConfigReader);
	DestroyFreeStorage (&Storage);
	return config;

}

int
WriteAudioOptions (const char *filename, char *myname, AudioConfig * config, unsigned long flags)
{
	ConfigDef    *AudioConfigWriter = NULL;
	FreeStorageElem *Storage = NULL, **tail = &Storage;
    register int i ;
	ConfigData cd ;

	if (config == NULL)
		return 1;
	cd.filename = filename ;

	if ((AudioConfigWriter = InitConfigWriter (myname, &AudioSyntax, CDT_Filename, cd)) == NULL)
		return 2;
	CopyFreeStorage (&Storage, config->more_stuff);
	/* building free storage here */
	/* playcmd */
	if( config->playcmd )
		tail = String2FreeStorage (&AudioSyntax, tail, config->playcmd, AUDIO_PLAYCMD_ID);
	/* delay */
	if( get_flags(config->set_flags, AUDIO_SET_DELAY) )
    	tail = Integer2FreeStorage (&AudioSyntax, tail, NULL, config->delay, AUDIO_DELAY_ID);
	/* rplay_host */
	if (get_flags(config->set_flags, AUDIO_SET_RPLAY_HOST) && config->rplay_host)
        tail = String2FreeStorage (&AudioSyntax, tail, config->rplay_host, AUDIO_RPLAY_HOST_ID);

	/* rplay_priority */
	if (get_flags(config->set_flags, AUDIO_SET_RPLAY_PRIORITY))
        tail = Integer2FreeStorage (&AudioSyntax, tail, NULL, config->rplay_priority, AUDIO_RPLAY_PRI_ID);

	/* rplay_volume */
	if( get_flags(config->set_flags, AUDIO_SET_RPLAY_VOLUME) )
        tail = Integer2FreeStorage (&AudioSyntax, tail, NULL, config->rplay_volume, AUDIO_RPLAY_VOL_ID);

	/* line structure */
    for( i = EVENT_ID_END-EVENT_ID_START-1 ; i >=0 ; i-- )
        if( config->sounds[i] )
        {
            FreeStorageElem **dtail = tail ;

            tail = Flag2FreeStorage (&AudioSyntax, tail, AUDIO_AUDIO_ID);
            if( *dtail )
                Path2FreeStorage (&AudioEventsSyntax, &((*dtail)->sub), NULL, config->sounds[i], EVENT_ID_START + i);
        }

    /* writing config into the file */
	WriteConfig (AudioConfigWriter, &Storage, CDT_Filename, &cd, flags);
	DestroyConfig (AudioConfigWriter);

	if (Storage)
	{
		fprintf (stderr, "\n%s:Config Writing warning: Not all Free Storage discarded! Trying again...", myname);
		DestroyFreeStorage (&Storage);
		fprintf (stderr, (Storage != NULL) ? " failed." : " success.");
	}
	return 0;
}
